moduleroot/CONTRIBUTING.md

```diff
@@ -186,32 +195,78 @@ This will now download a pre-fabricated image configured in the [default node-se
 install puppet, copy this module and install its dependencies per [spec/spec_helper_acceptance.rb](./spec/spec_helper_acceptance.rb)
 and then run all the tests under [spec/acceptance](./spec/acceptance).
 
-Writing Tests
--------------
+## Writing Tests
 
-XXX getting started writing tests.
+### Unit Tests
 
-If you have commit access to the repository
-===========================================
+When writing unit tests for Puppet, [rspec-puppet][] is your best friend. It provides tons of helper methods for testing your manifests against a 
+catalogue (e.g. contain_file, contain_package, with_params, etc). It would be ridiculous to try and top rspec-puppet's [documentation][rspec-puppet_docs] 
+but here's a tiny sample:
 
-Even if you have commit access to the repository, you will still need to
-go through the process above, and have someone else review and merge
-in your changes.  The rule is that all changes must be reviewed by a
-developer on the project (that did not write the code) to ensure that
-all changes go through a code review process.
+Sample manifest:
+```puppet
+file { "a test file":
+  ensure => present,
+  path   => "/etc/sample",
+}
+```
 
-Having someone other than the author of the topic branch recorded as
-performing the merge is the record that they performed the code
-review.
+Sample test:
+```ruby
+it 'does a thing' do
+  expect(subject).to contain_file("a test file").with({:path => "/etc/sample"})
+end
+```
 
+### Acceptance Tests
+
+Writing acceptance tests for Puppet involves [beaker][] and its cousin [beaker-rspec][]. A common pattern for acceptance tests is to create a test manifest, apply it
+twice to check for idempotency and/or errors, then run expectaions.
+
+```ruby
+it 'does an end-to-end thing' do
+  pp = <<-EOF
+    file { 'a test file': 
+      ensure  => present,
+      path    => "/etc/sample",
+      content => "test string",
+    }
+    
+  apply_manifest(pp, :catch_failures => true)
+  apply_manifest(pp, :catch_changes => true)
+  
+end
+
+describe file("/etc/sample") do
+  it { is_expected.to contain "test string" }
+end
 
-Additional Resources
-====================
+```
 
-* [Getting additional help](http://puppet.com/community/get-help)
+# If you have commit access to the repository
 
-* [Writing tests](https://docs.puppet.com/guides/module_guides/bgtm.html#step-three-module-testing)
+Even if you have commit access to the repository, you will still need to go through the process above, and have someone else review and merge
+in your changes.  The rule is that **all changes must be reviewed by a developer on the project (that did not write the code) to ensure that
+all changes go through a code review process.**
 
-* [General GitHub documentation](http://help.github.com/)
+Having someone other than the author of the topic branch recorded as performing the merge is the record that they performed the code
```

Comment:
Change:

Having someone other than the author of the topic branch recorded as performing the merge is the record that they performed the code

The record of someone performing the merge is the record that they performed the code review. Again, this must (should or whatever you want to use here) be someone other than the author of the topic branch.